File: src/PwrDrvr.MicroApps.DeployTool/S3Uploader.cs
using System;
using System.Data.Common;
using System.IO;
using System.Threading.Tasks;
using Amazon.S3;
using Amazon.S3.Transfer;

namespace PwrDrvr.MicroApps.DeployTool {
  internal class S3Uploader {
    private static readonly string _s3Bucket = "pwrdrvr-apps-staging";
    private static readonly string _tempDir = "./deploytool-temp";

    async internal static Task Upload(DeployConfig config) {
      var destinationPrefix = string.Format("{0}/{1}", config.AppName, config.SemVer);

      var s3Client = new AmazonS3Client();
      var s3TU = new TransferUtility(s3Client);

      // Make a local root dir for the upload
      string tempUploadPath = Path.Combine(_tempDir, destinationPrefix);
      if (Directory.Exists(tempUploadPath)) {
        Directory.Delete(tempUploadPath, true);
      }
      var tempUploadDir = Directory.CreateDirectory(tempUploadPath);

      // Copy the files in the source dir to the root dir
      // Note: It would be faster to move the files, then move them back
      DirectoryCopy(config.StaticAssetsPath, tempUploadDir.FullName, true);

      var diUploadRoot = new DirectoryInfo(_tempDir);

      // Do the upload
      await s3TU.UploadDirectoryAsync(diUploadRoot.FullName, _s3Bucket, "*.*", SearchOption.AllDirectories);

      // Delete the directory now that it's uploaded
      Directory.Delete(tempUploadPath, true);
    }

    // Really cheesy recursive function from:
    // https://docs.microsoft.com/en-us/dotnet/standard/io/how-to-copy-directories
    private static void DirectoryCopy(string sourceDirName, string destDirName, bool copySubDirs) {
      // Get the subdirectories for the specified directory.
      var dir = new DirectoryInfo(sourceDirName);

      if (!dir.Exists) {
        throw new DirectoryNotFoundException(
            "Source directory does not exist or could not be found: "
            + sourceDirName);
      }

      DirectoryInfo[] dirs = dir.GetDirectories();

      // If the destination directory doesn't exist, create it.       
      Directory.CreateDirectory(destDirName);

      // Get the files in the directory and copy them to the new location.
      FileInfo[] files = dir.GetFiles();
      foreach (FileInfo file in files) {
        string tempPath = Path.Combine(destDirName, file.Name);
        file.CopyTo(tempPath, false);
      }

      // If copying subdirectories, copy them and their contents to new location.
      if (copySubDirs) {
        foreach (DirectoryInfo subdir in dirs) {
          string tempPath = Path.Combine(destDirName, subdir.Name);
          DirectoryCopy(subdir.FullName, tempPath, copySubDirs);
        }
      }
    }
  }
}

File: src/microapps-publish/src/S3Uploader.ts
import path from 'path';
// eslint-disable-next-line import/no-extraneous-dependencies,import/no-unresolved
import { IDeployVersionPreflightResponse } from '@pwrdrvr/microapps-deployer';
import fs from 'fs-extra';
import { IConfig } from './config/Config';
import S3TransferUtility from './S3TransferUtility';

export default class S3Uploader {
  public static async Upload(
    config: IConfig,
    s3UploadPath: string,
    preflightResponse: IDeployVersionPreflightResponse,
  ): Promise<void> {
    try {
      //const destinationPrefix = `${config.app.name}/${config.app.semVer}`;

      // Parse the S3 Source URI
      const uri = new URL(s3UploadPath);
      const bucketName = uri.host;
      const destinationPrefix = uri.pathname.length >= 1 ? uri.pathname.slice(1) : '';

      // Make a local root dir for the upload
      const tempUploadPath = path.join(S3Uploader._tempDir, destinationPrefix);
      await S3Uploader.removeTempDirIfExists();
      await fs.mkdir(tempUploadPath, { recursive: true });

      // Copy the files in the source dir to the root dir
      // Note: It would be faster to move the files, then move them back
      // FIXME: Use p-map for controlled parallelism
      await fs.copy(config.app.staticAssetsPath, tempUploadPath);

      // Do the upload
      await S3TransferUtility.UploadDir(
        this._tempDir,
        destinationPrefix,
        bucketName,
        preflightResponse,
      );
    } finally {
      // Delete the directory, now that it's uploaded or if we failed
      await S3Uploader.removeTempDirIfExists();
    }
  }

  public static async removeTempDirIfExists(): Promise<void> {
    try {
      const stats = await fs.stat(S3Uploader._tempDir);
      if (stats.isDirectory()) {
        await fs.rm(S3Uploader._tempDir, { recursive: true });
      }
    } catch {
      // Don't care
      // fs.stat will throw if file/dir does not exist
      // Since we want the directory deleted this is ok
    }
  }

  private static readonly _tempDir = './deploytool-temp';
}
